### ISSUE TITLE
clink crashes when cd'ing to a long path.

### ISSUE THREAD

**Issue by jtnord** [chrisant996/clink#727]
When long paths are enabled https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry if you CD to a path exceeding the 255 characters clink hangs for a while and then crashes (0xc0000409).

e.g.

make a directory (in cmd.exe without clink running) `C:\source\temp\qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq\asdfasdfasdf\sdfasdfasdfasdfasdf\afsfasfasdfasdfasdfasdf`

in cmd.exe with without clink running cd to the directory

note that it works and you can run `dir` etc  (note explorer can not handle long paths, this is a long outstanding issue).

in a cmd.exe that has clink running cd to the same directory

note that it hangs for a while and then cmd.exe terminates (`[process exited with code 3221226505 (0xc0000409)`).

**Comment by chrisant996** [comment]
Are you using oh-my-posh for your prompt?

At first, I couldn't reproduce the crash.
I tried a bunch of different custom prompts.

And then I tried using oh-my-posh, and it crashed right away.

It looks like the Lua engine inside Clink is crashing, so the crash itself might be a Lua bug.

But it looks like a bug in oh-my-posh is triggering the crash.  Oh-my-posh's built-in Lua script assumes that launching oh-my-posh can never fail.  But when the current directory is longer than MAX_PATH, then oh-my-posh fails to launch, and then oh-my-posh's Lua script tries to dereference a `nil` table, and then Lua crashes.

```lua
local function run_posh_command(command)
    command = string.format('""%s" %s"', omp_executable, command)
    local _, is_main = coroutine.running()
    if is_main then
        return io.popen(command):read('*a')
--                              ^^^^^ This is incorrect. It must check whether io.popen() succeeded.
    end
    return io.popenyield(command):read('*a')
--                               ^^^^^ This is incorrect. It must check whether io.popen() succeeded.
end
```

I'll debug more tomorrow and find more specifics.
Maybe I'll be able to patch Lua.

But even if I can patch Lua, then the oh-my-posh script will throw an error and force Lua to print an ugly error message.
Ultimately, this seems to be a bug in oh-my-posh itself, and needs to be fixed in oh-my-posh.

https://github.com/JanDeDobbeleer/oh-my-posh/blob/a66586da6936803367bcf9c86e784b7cb92925d2/src/shell/scripts/omp.lua#L88-L90

**Comment by jtnord** [comment]
> Are you using oh-my-posh for your prompt?

no, but I am using flexprompt `flexprompt.settings.left_prompt = "{battery}{histlabel}{cwd:shorten=smartroot}{k8s}{git}"
`

**Comment by jtnord** [comment]
> The crash is happening inside the C runtime function _popen() itself.

possibly the working directory of it needs to be "longified" with `\\?\`  

Dumpfile sent.

**Comment by jtnord** [comment]
this may be my lua that is triggering the crash :-( 
https://github.com/jtnord/flex-prompts/blob/2767b185b28462881637c852a2c206b17b27aa86/flexprompt_k8s.lua#L7 

attempting to start `kubectl` in the command prompt returns `The current directory is invalid`

**Comment by chrisant996** [comment]
> this may be my lua that is triggering the crash :-( https://github.com/jtnord/flex-prompts/blob/2767b185b28462881637c852a2c206b17b27aa86/flexprompt_k8s.lua#L7
> 
> attempting to start `kubectl` in the command prompt returns `The current directory is invalid`

It turns out that although yes, that code in flexprompt_k8s.lua is missing an error check, the crash is happening inside the `io.popen()` API itself, before it returns an error.

There is a bug in the C runtime code for `_popen()`, which is the underlying implementation for the Lua `io.popen()` API.

The bug is inside the `common_popen_nolock()` function in the conio\popen.cpp file in the C runtime sources (which are included with Visual Studio).

```cpp
    // Associate a stream with the pipe handle to be returned to the caller:
    unique_stream pipe_stream(stdio_traits::tfdopen(pipe_handles[1], fdopen_mode));
```

The `pipe_stream` variable now contains a copy _by value_ of the output handle from `pipe_handles`.  At this point, **BOTH** `pipe_stream` and the caller who passed in `pipe_handles` each think they own the responsibility to free the handle.

That can be ok as a temporary state, as long as one of them relinquishes control to the other soon afterwards...  But that's precisely what goes wrong -- the `popen_common_nolock()` function has a bug in its error handling:

There are 4 places where it can exit without `pipe_handles` relinquishing control of the handle.
It only relinquishes control upon _successful_ completion.
In error cases, it closes the handle, even though its caller really owns the responsibility for closing the handle.

```cpp
    PROCESS_INFORMATION process_info = PROCESS_INFORMATION();
    BOOL const child_status = stdio_traits::create_process(
        selected_cmd_exe,
        command_line.get(),
        nullptr,
        nullptr,
        TRUE,
        0,
        nullptr,
        nullptr,
        &startup_info,
        &process_info);

    if (!child_status)
        return nullptr;

    FILE* const result_stream = pipe_stream.detach();
```

The line `if (!child_status)` detects the failure and returns.
But notice that `pipe_stream` doesn't `detach()` until _after_ checking for failure.
It only relinquishes responsibility for closing the handle upon success.

So, upon error, `pipe_stream` is still holding on to the handle, so it closes the handle before returning.

And then the caller, `common_popen()`, proceeds to also close the handle because of the failure:

```cpp
        errno_t const saved_errno = errno;

        return_value = common_popen_nolock(
            command,
            fdopen_mode.mode,
            std_fh,
            ordered_pipe_handles);

        errno = saved_errno;

        if (return_value != nullptr)
            __leave;

        // If the implementation function returned successfully, everything was
        // cleaned up except the lock.
        int* const first = ordered_pipe_handles;
        int* const last  = first + _countof(ordered_pipe_handles);
        for (int* it = first; it != last; ++it)
        {
            if (*it != -1)
                _close(*it);
        }
```

When `return_value` is nullptr (failure), then `common_popen()` closes the handles.

There are two ways to look at this:
1. `common_popen_nolock()` could have fully taken ownership and set _both_ the in and out handles to `-1`.
2. `common_popen_nolock()` could have not closed the handle that it didn't reset to `-1`.

Anyway, the `_close()` in `common_popen()` causes a crash, because `common_popen_nolock()` already closed the handle, and it's invalid to close a handle twice (something else could have already reused that handle ID).  If it didn't force a crash then it could lead to malfunctions and security flaws later on.

So any failure will cause a crash.

There's nothing the calling program can really do about that, other than "don't use `_popen()`".  But that isn't very realistic.

Fortunately, Clink already had to make its own custom implementation of `io.popenyield()` which has a custom alternative implementation of `_popen()`, so I'm very familiar with the internals of how to write a `_popen()`-like function.

I will need to replace the C runtime `_popen()` function with a custom version that doesn't have that bug.

In fairness, that's a difficult bug to catch in the C runtime.  Because `_popen()` _always_ actually launches `%SystemRoot%\System32\cmd.exe /c {the_command_line_passed_to_popen}`.  And it's pathologically rare for launching cmd.exe to fail -- there are very few ways for that to fail, but this case is one of the very few ways that could happen.

**Comment by chrisant996** [comment]
Fixing this isn't quick or easy -- I have to make a custom implementation of a C runtime function, and replace all usage of that C runtime function.

But, fortunately, it's straightforward how to do that, and I've done it before for other C runtime functions.

So, I'll add a workaround to avoid the crash, but it could be several days (or weeks, depending on how much free time I have to spend on this).

In the meantime, if you update your flexprompt_k8s.lua script to model it after the built-in k8s module and use async prompt filtering, then you can avoid the crash, and also speed up your prompt.

**Comment by chrisant996** [comment]
There's a much easier workaround for the C runtime bug.

I can just use [_set_invalid_parameter_handler()](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/set-invalid-parameter-handler-set-thread-local-invalid-parameter-handler?view=msvc-170) to tell the C runtime to gracefully return an error instead of forcibly terminating the process.

And that's a _**much**_ better approach anyway, because it turns out Lua APIs often pass inputs directly down in the C runtime, expecting the C runtime to validate the inputs.  In other words, any Lua script can easily force a crash just by, for example, passing a mode string with an unexpected character to a file IO API.

For example, `io.open("somefile", "Qrwt")`.  The `Q` is an invalid mode character, and the C runtime will by default terminate the process out of an abundance of caution, in case the caller forgot to provide error handling code.  That's unrealistic and unnecessary in the context Clink or other programs which hosts Lua.

It's still quite interesting that the C runtime itself has a double-close bug inside `_popen()`, though!  It does detect that it's about to do something invalid, though, and it blocks itself from _actually_ performing a double-close.  So, while it's definitely a bug, there is no risk of harm, apart from the fact that the C runtime kills the process as a "safest default behavior" mitigation.

**Comment by chrisant996** [comment]
### Perfect, that workaround works.

But then one of four things happens, depending on how the prompt script is implemented:

## 1. If a prompt filter doesn't test whether `io.popen()` failed...

Then Lua will report an error such as the following:

```plain
prompt filter failed:
[string "---@diagnostic disable: undefined-global..."]:83: attempt to index a nil value
stack traceback:
        [string "---@diagnostic disable: undefined-global..."]:195: in function 'func'
        ...s/clink/.build/vs2022/../../clink/app/scripts/prompt.lua:137: in function <...s/clink/.build/vs2022/../../cli
nk/app/scripts/prompt.lua:124>
        [C]: in function 'xpcall'
        ...s/clink/.build/vs2022/../../clink/app/scripts/prompt.lua:191: in function <...s/clink/.build/vs2022/../../cli
nk/app/scripts/prompt.lua:105>
        (...tail calls...)


ERROR IN LUA PROMPT FILTER
c:\wbin\clink\abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzab
cdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqr
stuvwxyz1234567890123456>
```

## 2. If a prompt filter just uses the `io.popen()` call to add information to a prompt...

Then the prompt will simply behave as though it couldn't the information, and the prompt will work, but won't have the additional information.

This is how flexprompt and the built-in clinkprompts behave.

## 3. If a prompt filter tests for failure and gives up...

Then the prompt will be completely empty.
No prompt.
Like the following screen shot.
Which is confusing, but is _correct behavior_ given that the prompt script chose to not provide any alternative prompt text upon failure -- i.e. Clink is simply doing what the script told Clink to do.

![Image](https://github.com/user-attachments/assets/97cddda6-cd80-4638-aba2-09480832af9b)

## 4. If a prompt filter tests for failure and returns a placeholder default prompt string...

Then that's what will be displayed.
But I don't know of any prompt filters that do this.

**Comment by jtnord** [comment]
Thanks for your detailed investigation and analysis here, it is very insightful!

> In the meantime, if you update your flexprompt_k8s.lua script to model it after the built-in k8s module and use async prompt filtering, then you can avoid the crash, and also speed up your prompt.

> Calling io.popenyield() in the main coroutine is nonsensical

now that there is support in flexprompt I think I will be better upgrading and removing my code :)

**Comment by chrisant996** [comment]
This was a great find, thanks for reporting it!  I had no idea that the `_popen()` function in the C runtime had this bug (at least Visual Studio's C runtime).  This also led to realizing that `io.popenyield()` accidentally missed returning the usual values upon failure (nil, error message, error code).
